core/src/storage2/alloc/init.rs

```diff
@@ -66,47 +75,127 @@ impl From<ContractPhase> for DynamicAllocatorState {
+                *self = DynamicAllocatorState::Finalized;
+            }
+            DynamicAllocatorState::Finalized => {
+                panic!(
+                    "cannot finalize the dynamic storage allocator \
+                 after it has already been finalized"
+                )
+            }
+            DynamicAllocatorState::UninitCall | DynamicAllocatorState::UninitDeploy => {
+                // Nothing to do in these states.
+            }
+        }
+    }
+
+    /// Runs the closure on the global instance for the dynamic storage allocator.
+    ///
+    /// Will automatically initialize the global allocator instance if it has not
+    /// yet been initialized.
+    ///
+    /// # Panics
+    ///
+    /// If the global dynamic storage allocator instance has already been finalized.
+    pub fn on_instance<F, R>(&mut self, f: F) -> R
+    where
+        F: FnOnce(&mut DynamicAllocator) -> R,
+    {
+        match self {
+            DynamicAllocatorState::UninitDeploy => {
+                let mut allocator = DynamicAllocator::default();
+                let result = f(&mut allocator);
+                *self = DynamicAllocatorState::Initialized(allocator);
+                result
+            }
+            DynamicAllocatorState::UninitCall => {
+                let mut allocator = pull_spread_root::<DynamicAllocator>(&Key(
+                    DYNAMIC_ALLOCATOR_KEY_OFFSET,
+                ));
+                let result = f(&mut allocator);
+                *self = DynamicAllocatorState::Initialized(allocator);
+                result
+            }
+            DynamicAllocatorState::Initialized(ref mut allocator) => f(allocator),
+            DynamicAllocatorState::Finalized => {
+                panic!(
+                    "cannot operate on the dynamic storage \
+                     allocator after it has been finalized"
+                );
+            }
+        }
+    }
+}
+
 cfg_if! {
     if #[cfg(all(not(feature = "std"), target_arch = "wasm32"))] {
         // Procedures for the Wasm compilation:
 
         /// The global instance for the dynamic storage allocator.
         static mut GLOBAL_INSTANCE: DynamicAllocatorState = DynamicAllocatorState::UninitDeploy;
 
-        /// Commands the (re-)initialization of the global instance for the dynamic
-        /// storage allocator.
-        pub fn initialize_for(phase: ContractPhase) {
-            let instance = unsafe { &mut GLOBAL_INSTANCE };
-            // We do not allow reinitialization for Wasm targets for performance reasons.
-            if let DynamicAllocatorState::Initialized(_) = instance {
-                panic!("cannot reinitialize dynamic storage allocator instance in Wasm");
-            }
-            *instance = phase.into();
+        /// Forwards to the `initialize` of the global dynamic storage allocator instance.
+        pub fn initialize(phase: ContractPhase) {
+            unsafe { &mut GLOBAL_INSTANCE }.initialize(phase);
```

Comment:
Prove or remove ;) Would be nice to have some reasoning why this is safe. The single threaded aspect might be obvious but I am not so sure about reentrancy on the top of my head.

Reply:
Good catch! Will add a comment there.